### DISCUSSION TITLE
Next.js rewrites work, but Javascript and CSS is 404

### DISCUSSION THREAD

**Discussion by leo-petrucci** [vercel/next.js#51797] [Help] [answered]
### Summary

I'm trying to run multiple Vercel/Next.js projects under the same domain as [suggested on this guide][1].

All of my different apps have very different requirements so, it makes sense for me to separate them into their own projects.

I have a main Next.js project which will run at `mysite.com`. This project contains rewrites to the other projects:

```
const rewrites = async () => [
  {
    source: "/topic/",
    destination: "https://forums.mysite.com/topic/",
  },
  {
    source: "/topic/:match*",
    destination: "https://forums.mysite.com/topic/:match*",
  },
];

/**
 * @type {import('next').NextConfig}
 */
const nextConfig = {
  rewrites,
};

module.exports = nextConfig;
```

These rewrites are working, kind of. If I visit `mysite.com/topic/` Next.js attempt to load the site at `forums.mysite.com`, however it never displays because all Javascript scripts go to 404.

It's essentially attempting to load `forums.mysite.com`'s scripts on `mysite.com`'s domain, where they obviously don't exist.

Is there any way around this? Seems like a pretty big oversight but this is definitely [the way that Next.js suggests this is done][2].

Any ideas on how I can fix this?

The issue is obviously related to Next.js loading scripts as absolute paths (e.g. `/_next/script/etc...`) instead of using the full domain (e.g. `https://mysite.com/_next/script/etc...`). But I don't think the option to do that exists?

  [1]: https://vercel.com/guides/how-can-i-serve-multiple-projects-under-a-single-domain
  [2]: https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites#incremental-adoption-of-nextjs

**Comment by rasif-sahl** [comment, marked as answer]
To resolve the issue with scripts not loading correctly when using rewrites in Next.js, you can configure Next.js to use the full domain when loading the scripts. You can achieve this by setting the assetPrefix configuration option in your Next.js project.

Here's how you can modify your code to fix the issue:

In your next.config.js file, add the assetPrefix configuration option and set it to the full domain of your project, including the protocol (e.g., https://mysite.com):

```
const nextConfig = {
  rewrites,
  assetPrefix: 'https://mysite.com',
};

module.exports = nextConfig;
```

Deploy your Next.js project to ensure the changes take effect.
By setting the assetPrefix option, Next.js will prepend the specified URL to the asset paths (e.g., scripts, stylesheets) when generating the HTML. This ensures that the correct full domain is used when loading the scripts, addressing the issue you encountered.

Please note that when using the assetPrefix option, you need to ensure that your assets (scripts, stylesheets, etc.) are accessible from the specified full domain (https://mysite.com in this case). Make sure that your server is correctly configured to serve the assets from the appropriate location.

By configuring the assetPrefix option, you can make Next.js use the full domain when loading scripts, resolving the issue of scripts not being found when using rewrites in your project.

**Reply by nathanmrtns** [reply]
This fixes for two next js applications. Is there a way to do it for non nextjs "destination"?

**Reply by leo-petrucci** [reply]
@nathanmrtns I assume you'd have to change the destination's scripts so they're not relative.

**Reply by alexcastrodev** [reply]
@leo-petrucci I have this problem also, and i opened a [issue](https://forum.ghost.org/t/using-ghost-pro-with-a-subdirectory-nuxtjs/41171/7) on Ghost Forum. 

But, reading this, i wonder: by changing assetPrefix will change not only the route, but the whole App, innit ?

**Reply by alexcastrodev** [reply]
Also seens a workaround, because this is used for CDN as types says.

**Reply by leo-petrucci** [reply]
> @leo-petrucci I have this problem also, and i opened a [issue](https://forum.ghost.org/t/using-ghost-pro-with-a-subdirectory-nuxtjs/41171/7) on Ghost Forum. 
> 
> But, reading this, i wonder: by changing assetPrefix will change not only the route, but the whole App, innit ?

No, asset prefix just changes the prefix of the assets 👍

**Reply by alexcastrodev** [reply]
Exacly, if you are trying to use a website as a subdirectory, i don't want to change all of my assets to this destination.
For example, if i have /forum ( with ipboard ) and /blog ( Nuxt ) and they have differents prefix, it will breaks.

In my case, we have our website, but i want only /blog rewrites and all /blog/:path has his own assetPrefix. not for whole application assets.

**Reply by leo-petrucci** [reply]
> Exacly, if you are trying to use a website as a subdirectory, i don't want to change all of my assets to this destination.
> For example, if i have /forum ( with ipboard ) and /blog ( Nuxt ) and they have differents prefix, it will breaks.
> 
> In my case, we have our website, but i want only /blog rewrites and all /blog/:path has his own assetPrefix. not for whole application assets.

Funnily enough I'm also doing something very similar with ipboard.

My project is a monorepo.

I have a central app which handles all the rewrites, then I have a /blog app, which the central app has rewrites to display under the main domain.

In my /blog project I have rewrites to make sure all assets work correctly.

**Reply by alexcastrodev** [reply]
lol, i miss to colaborate on IPBoard Forums. 

But here, in my case, i dont have access to /blog code, because it's a Ghost App managed.

**Reply by leo-petrucci** [reply]
> But here, in my case, i dont have access to /blog code, because it's a Ghost App managed.

Ah that sucks, unless there's a setting on Ghost to prefix all assets then I can't really think of a way it could work.

**Reply by alexcastrodev** [reply]
Nah, it will also change others rewrites routes. I've stuck. lol

May i create an issue ?

**Reply by CamiloERH** [reply]
did u solve it, facing the same issue D:

**Reply by alexcastrodev** [reply]
Nothing …

**Reply by CamiloERH** [reply]
still testing but i think this works for me I was not using the fallback option as it says in

https://nextjs.org/docs/app/api-reference/next-config-js/rewrites#incremental-adoption-of-nextjs

module.exports = {
  async rewrites() {
    return {
      fallback: [
        {
          source: '/:path*',
          destination: `https://custom-routes-proxying-endpoint.vercel.app/:path*`,
        },
      ],
    }
  },
}

**Comment by stormi99** [comment]
Hey any updates?

**Reply by leo-petrucci** [reply]
What do you mean? It's fixed.

**Reply by anthony-brown-5pm** [reply]
It is? Still having this issue today. Can't load css/js from my other nextjs app from rewrites.

**Comment by leo-petrucci** [comment]
> It is? Still having this issue today. Can't load css/js from my other nextjs app from rewrites.

Oh, you might have to add those rewrites in `beforeFiles`:
```
    beforeFiles: [
      // for the server list
      {
        source: `/_next/data/:release/my-app.json`,
        destination: `${process.env.NEXT_MY_APP_URL}/_next/data/:release/my-app.json`,
      },
      {
        source: `/_next/data/:release/my-app/:path*`,
        destination: `${process.env.NEXT_MY_APP_URL}/_next/data/:release/my-app/:path*`,
      },
    ]
```
